clientapi/routing/register_test.go

```diff
@@ -208,3 +212,123 @@ func TestValidationOfApplicationServices(t *testing.T) {
 		t.Errorf("user_id should not have been valid: @_something_else:localhost")
 	}
 }
+
+func TestValidatePassword(t *testing.T) {
+	t.Parallel()
+	defaults := &config.PasswordRequirements{}
+	defaults.Defaults()
+
+	custom := &config.PasswordRequirements{
+		MinPasswordLength: 16,
+		MaxPasswordLength: 32,
+		RequireMixedCase:  true,
+		MinNumberSymbols:  5,
+	}
+	var testCases = []struct {
+		name     string
+		config   config.PasswordRequirements
+		password string
+		expected *util.JSONResponse
+	}{
+		{
+			"default reject too short",
+			*defaults,
+			"foobar",
+			&util.JSONResponse{
+				Code: http.StatusBadRequest,
+				JSON: jsonerror.WeakPassword(fmt.Sprintf("password too weak: min %d chars", defaults.MinPasswordLength)),
+			},
+		},
+		{
+			"default reject too long",
+			*defaults,
+			// len 600
+			"aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa",
+			&util.JSONResponse{
+				Code: http.StatusBadRequest,
+				JSON: jsonerror.WeakPassword(fmt.Sprintf("'password' >%d characters", defaults.MaxPasswordLength)),
+			},
+		},
+		{
+			"default allow long enough",
+			*defaults,
+			"thisisalongenoughpassword",
```

Comment:
This is brittle. If we change the default length of passwords then this test will fail because we're assuming the values. I see little benefit in testing this. By all means, check that:
```go
defaults := &config.PasswordRequirements{}
defaults.Defaults()
```
sets some sensible values (e.g min > 0 and max > min) but try to not bake in assumptions about the values.

Reply:
removed this test. added a test in `setup/config/config_clientapi_test.go` to make sure the values are set correctly.